# Sultan

Group: Final_Project_Group 1 | Workshop: Performance Metrics Classification (MNIST)

Section: cross_validate, accuracy, DummyClassifier, cross-validation folds

Helper class for this section: `CrossValReport` in `cross_val_report.py`. Helper code lives in the class as methods; this notebook only imports and calls it.

In [1]:
from src.cross_val_report import CrossValReport

report = CrossValReport(cv=3).load_data().run()
report.results_table()

,fold,SGDClassifier,DummyClassifier
0,Fold 1,0.95035,0.90965
1,Fold 2,0.96035,0.90965
2,Fold 3,0.96040,0.90965


## Sultan

### cross_validate

**Question:** Describe in your own words what the `cv` argument is for.

**Answer:**

cv tells the model how many parts the training data should be divided into for cross-validation. With cv=3, the data is split into 3 folds. The model trains on 2 folds and tests on the remaining fold. This happens 3 times so that each fold gets used as the test set once.

**Question:** How many times has the model been trained?

**Answer:**
The model has been trained 3 times, once for each fold. Since there are 60,000 training images, each time it uses 40,000 images for training and 20,000 images for testing.

**Question:** How is the value of `cv` related to the output of this function?

**Answer:**

The number of folds determines how many results we get. Since cv=3, fit_time, score_time, and test_score each have 3 values, one for each fold.

**Question:** What do `fit_time`, `score_time` and `test_score` stand for?

**Answer:**

- fit_time is how long the model takes to train for each fold.
- score_time is how long it takes to make predictions and calculate the score.
- test_score is the accuracy of the model on the test data for each fold.

### Accuracy

**Question:** Provide a textual description of the accuracy result for the first fold, e.g. "out of 100 digits, XYZ were correctly predicted, while ZYX were incorrectly predicted".

**Answer:**

In the first test fold, there were 20,000 digits. The model correctly predicted 19,007 of them and got 993 wrong, giving an accuracy of 95.04%.

The calculation is:

0.95035 × 20,000 = 19,007

In [2]:
print(report.describe_fold(fold=0))

Fold 1: out of 20,000 digits, 19,007 were correctly predicted, while 993 were incorrectly predicted (accuracy = 95.03%).


**Question:** Given the uniform distribution of samples across all 10 digits, and that the classifier classifies one of them versus all the others (`5` versus all others), what would be a simple handwritten rule to achieve 90% accuracy?

**Answer:**

A simple rule would be to always predict that the digit is not a 5. Since around 10% of the digits are 5s, this would still give about 90% accuracy. The problem is that the rule would never actually detect a 5.

**Question:** How would you describe the goodness of classification of the `SGDClassifier` versus the expected accuracy of such a handwritten rule? (a sentence or two)

**Answer:**

The SGDClassifier gets around 95.7% accuracy on average, while the simple rule gets around 90%. So the SGDClassifier is clearly better, although the improvement is not as big as it might seem because the simple rule already gets around 90% by just ignoring all the 5s.

### DummyClassifier

**Question:** What is the accuracy of the `DummyClassifier`?

**Answer:**

The DummyClassifier gets about 90.965% accuracy on each fold. This is because it always predicts "not 5", and around 9% of the digits in the dataset are 5s.

In [3]:
print(report.dummy_results["test_score"])

[0.90965 0.90965 0.90965]


**Question:** How does it compare to the handwritten rule you suggested above, and to the `SGDClassifier`?

**Answer:**

The DummyClassifier gives almost the same result as my simple handwritten rule, around 90–91%. The SGDClassifier is about 5 percentage points higher, with accuracy around 95–96%.

This shows why accuracy by itself can be misleading. A model can get a high accuracy score without actually detecting any 5s. Looking at precision and recall gives us a better idea of how well the SGDClassifier is actually detecting 5s. Its precision is 0.84 and its recall is 0.65.

**Visual:** Grouped bar chart of accuracy per fold, SGDClassifier next to DummyClassifier.

In [4]:
report.plot_accuracy_per_fold()

### Cross-validation folds (research prompt)

**Question:** Research what a cross-validation fold is and explain it in your own words.

**Answer:**

A cross-validation fold is basically one part of the dataset. The dataset is divided into k parts, and each part gets a turn being used for testing while the other parts are used for training.

This means every sample gets tested once. At the end, the scores from all the folds can be averaged to get a more reliable idea of how the model performs.

For cv=3, the 60,000 images are divided into 3 folds, so each fold contains 20,000 images.